src/rational/q/arithmetic/sub.rs

```diff
@@ -108,3 +150,178 @@ mod test_sub {
         );
     }
 }
+
+#[cfg(test)]
+mod test_sub_between_z_and_q {
+
+    use super::Z;
+    use crate::rational::Q;
+    use std::str::FromStr;
+
+    /// testing subtraction for [`Q`] and [`Z`]
+    #[test]
+    fn sub() {
+        let a: Q = Q::from_str("5/7").unwrap();
+        let b: Z = Z::from(4);
+        let c: Q = a - b;
+        assert_eq!(c, Q::from_str("-23/7").unwrap());
+    }
+
+    /// testing subtraction for both borrowed [`Q`] and [`Z`]
+    #[test]
+    fn sub_borrow() {
+        let a: Q = Q::from_str("5/7").unwrap();
+        let b: Z = Z::from(4);
+        let c: Q = &a - &b;
+        assert_eq!(c, Q::from_str("-23/7").unwrap());
+    }
+
+    /// testing subtraction for borrowed [`Q`] and [`Z`]
+    #[test]
+    fn sub_first_borrowed() {
+        let a: Q = Q::from_str("5/7").unwrap();
+        let b: Z = Z::from(4);
+        let c: Q = &a - b;
+        assert_eq!(c, Q::from_str("-23/7").unwrap());
+    }
+
+    /// testing subtraction for [`Q`] and borrowed [`Z`]
+    #[test]
+    fn sub_second_borrowed() {
+        let a: Q = Q::from_str("5/7").unwrap();
+        let b: Z = Z::from(4);
+        let c: Q = a - &b;
+        assert_eq!(c, Q::from_str("-23/7").unwrap());
+    }
+
+    /// testing subtraction for big numbers
+    #[test]
+    fn sub_large_numbers() {
+        let a: Q = Q::from_str(&format!("{}/2", u64::MAX)).unwrap();
+        let b: Q = Q::from_str(&format!("1/{}", u64::MAX)).unwrap();
+        let c: Z = Z::from(u64::MAX);
+        let d: Q = a - &c;
+        let e: Q = b - c;
+        assert_eq!(
+            d,
+            Q::from_str(&format!("{}/2", u64::MAX)).unwrap()
+                - Q::from_str(&format!("{}/1", u64::MAX)).unwrap()
+        );
+        assert_eq!(
+            e,
+            Q::from_str(&format!("1/{}", u64::MAX)).unwrap()
+                - Q::from_str(&format!("{}/1", u64::MAX)).unwrap()
+        );
+    }
+}
+
+#[cfg(test)]
+mod test_sub_between_types {
+
+    use crate::rational::Q;
+
+    /// testing subtraction between different types
+    #[test]
+    #[allow(clippy::op_ref)]
+    fn sub() {
+        let a: Q = Q::from(42);
+        let b: u64 = 1;
+        let c: u32 = 1;
+        let d: u16 = 1;
+        let e: u8 = 1;
+        let f: i64 = 1;
+        let g: i32 = 1;
+        let h: i16 = 1;
+        let i: i8 = 1;
+        let j: f32 = 1.3;
+        let k: f64 = 1.3;
```

Comment:
The variable names don't really explain what the value is about.
Maybe rename `a` to something like `rational` or `q_value`.
Similarly:
`b` -> `unsinged_int_64` or `uint_64`
`f` -> `int_64`
`j` -> `float_32`
etc.

Reply:
discussed in group and current state is accepted